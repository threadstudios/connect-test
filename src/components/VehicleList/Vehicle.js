import Inferno from 'inferno';
import fetch from 'isomorphic-fetch';
import Loading from '../Loading/Loading';
import ErrorComponent from '../Error/Error';
import Image from '../PreloadedImage';
import {compose, branch, withLifecycle, withState, defaultProps, renderComponent} from 'incompose';

const Vehicle = (props) => {
    return (
        <div className={props.className}>
            <Image className="vehicle_image" url={props.media[0].url} />
            <div className="vehicle_description">
                <h1><span>{props.id.toUpperCase()}</span></h1>
                <h2>From {props.data.price}</h2>
                <p>{props.data.description}</p>
            </div>
        </div>
    );
}

const fetchData = (id) => {
    const apiUrl = process.env.API_URL ? process.env.API_URL : __CONFIG__.API_URL;
    return new Promise((resolve, reject) => {
        fetch(`${apiUrl}api/vehicle/${id}`)
        .then((result) => {
            return result.json();
        })
        .then((json) => {
            return resolve(json);
        })
        .catch((err) => {
            return reject(err);
        });
    })
}

const withLoadingComponent = branch(
    (props) => props.loading,
    renderComponent(Loading)
);

const withErrorComponent = branch(
    (props) => props.error,
    renderComponent(ErrorComponent)
)

const withFetcher = branch(
    (props) => !props.data,
    withLifecycle({
        componentDidMount() {
            this.setState({loading: true});
            fetchData(this.props.id)
            .then((result) => {
                this.setState({
                    loading: false,
                    data : result
                })
            })
            .catch((err) => {
                this.setState({
                    loading: false,
                    error: true,
                    msg: err.message
                })
            });
        }
    })
)   

export default compose(
    withState('data', 'setData', false),
    withState('loading', 'setLoading', true),
    withState('error', 'setError', false),
    withState('msg', 'setErrorMsg', false),
    defaultProps({ className : 'vehicle' }),
    withFetcher,
    withLoadingComponent,
    withErrorComponent
)(Vehicle)