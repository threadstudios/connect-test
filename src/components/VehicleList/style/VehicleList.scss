$border : 1px solid #EDEDED;

.vehicle_list {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
}

.vehicle {
    width: 100%;
    display: flex;
    flex-direction: row;
    min-height: 25%;
    @include breakpoint($medium) {
        width: 50%;
        flex-wrap: wrap;
        align-items: stretch;
        justify-content: space-around;
        border-bottom: $border;
        align-content: flex-start;
        .vehicle_description {
            border-right: $border;
        }
        &:nth-child(even) .vehicle_description {
            border-right: none;
        }
    }
    @include breakpoint($large) {
        width: 25%;
        &:nth-child(1n) .vehicle_description {
            border-right: $border;
        }
        &:nth-child(4n) .vehicle_description {
            border-right: none;
        }
    }
}

.vehicle_image {
    width: 22%;
    min-width: 22%;
    height: auto;
    text-align: center;
    @include breakpoint($medium) {
        width: 100%;
        min-width: 100%;
        height: 33%;
    }
    background-size: cover;
    background-repeat: no-repeat;
    background-position: 50% 50%;
}

.vehicle_description {
    border-bottom: $border;
    text-align: left;
    padding: 16px 24px;
    @include breakpoint($medium) {
        text-align: center;
        border-bottom: none;
        border-right: none;
        width: 100%;
    }
}