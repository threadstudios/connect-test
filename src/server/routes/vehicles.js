import { Router } from 'express';
import vehicles from 'technical-test/server_api/vehicles';

const api = new Router();

api.use((req, res, next) => {
    req.store = vehicles;
    return next();
});

api.get('/api/vehicle', (req, res) => {
    return res.json(vehicles);
});

api.get('/api/vehicle/:id', (req, res) => {
    // clean the id
    let id = req.params.id.replace(/[^\w\s\-]*/, "");
    try {
        let vehicle = require(`technical-test/server_api/vehicle_${id}`);
        return res.json(vehicle);
    } catch(e) {
        return res.status(400).json({
            error: 'Vehicle could not be found'
        })
    }
});

export default api;