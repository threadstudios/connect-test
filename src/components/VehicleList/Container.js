import Inferno from 'inferno';
import { compose, withLifeCycle, defaultProps } from 'incompose';
import Vehicle from './Vehicle';

const List = ({vehicles}) => {
    return (
        <div className="vehicle_list">
            {vehicles.map((vehicle) => {
                return <Vehicle {...vehicle} />
            })}
        </div>
    );
}

export default compose(
    defaultProps({
        vehicles: []
    })
)(List);