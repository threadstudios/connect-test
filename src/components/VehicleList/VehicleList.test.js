import Inferno from 'inferno';
import { renderToSnapshot } from 'inferno-test-utils';
import VehicleList from './Container';
import vehicles from 'technical-test/server_api/vehicles';

describe("A Vehicle List Component", () => {
    
    window.__CONFIG__ = {API_URL : '/'}

    it("should render without an error", () => {
        expect(renderToSnapshot(<VehicleList {...vehicles} />)).toMatchSnapshot();
    })

})