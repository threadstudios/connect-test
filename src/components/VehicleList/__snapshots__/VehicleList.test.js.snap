// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A Vehicle List Component should render without an error 1`] = `
<div>
  <div
    id="xe"
    media={
      Array [
        Object {
          "name": "vehicle",
          "url": "/images/xe_k17.jpg",
        },
      ]
    }
    modelYear="k17"
    url="/api/vehicle/xe"
  />
  <div
    id="xf"
    media={
      Array [
        Object {
          "name": "vehicle",
          "url": "/images/xf_k17.jpg",
        },
      ]
    }
    modelYear="k17"
    url="/api/vehicle/xf"
  />
  <div
    id="xj"
    media={
      Array [
        Object {
          "name": "vehicle",
          "url": "/images/xj_k16.jpg",
        },
      ]
    }
    modelYear="k16"
    url="/api/vehicle/xj"
  />
  <div
    id="fpace"
    media={
      Array [
        Object {
          "name": "vehicle",
          "url": "/images/fpace_k17.jpg",
        },
      ]
    }
    modelYear="k17"
    url="/api/vehicle/fpace"
  />
  <div
    id="ftype"
    media={
      Array [
        Object {
          "name": "vehicle",
          "url": "/images/ftype_k17.jpg",
        },
      ]
    }
    modelYear="k17"
    url="/api/vehicle/ftype"
  />
</div>
`;
